### PR TITLE
GPU: DisplayTransfer & MemoryFill cleanup and param check

### LINKED ISSUE #2014: Problematic implementation of (not accelerated) DisplayTransfer
This is a really old and known one... just open this issue so that we won't forget.

----

https://github.com/citra-emu/citra/blob/7b4dcacbb2006de6483e982b21956a8f3098aa1d/src/core/hw/gpu.cpp#L267

y, as a loop variable, is modified in the loop, which looks no way to be right.

### PR DESCRIPTION
**Pull request by wwylele**
Sorry for putting many things into one PR. But there are so many issues at the same place, so I prefer working on them together.

This PR should be reviewed commit by commit:

 - `rasterizer: separate TextureCopy from DisplayTransfer` & `gpu: move MemoryFill, TextureCopy and DisplayTransfer into functions`: the original code indented too much to read. Split them into different functions

 - `gpu: add validity check for TextureCopy, DisplayTransfer and FillMemory`: an alternative to #2078 & #2057. IMO all validity check should be performed before dispatching to rasterizer. I still didn't do hardware test for these though. Hardware tests can be done after this cleanup.  

   -  `memory: fix IsValidVirtualAddress for RasterizerCachedMemory`: bonus from the commit above

 -  ` gpu: keep the old signal strategy for null pointer`: due to previous commits, the interrupt behaves differently from before, which seems causing some logic error in games (only "unmapped Read/Write" in log, no actual regression in game play, though). This commit regressed the interrupt to the behaviour same as the one in master. Still needs further testing in the future.

 - `gpu: DisplayTransfer: a less amazing algorithm for flip`: close #2014. (Quiz: old implementation actually works for most cases! can you figure it out?)

### CODE REVIEW

**Comment by lioncash** [comment]
You should probably be including your extended descriptions that are in the initial PR message inside the actual commit messages themselves (and expanding on them where necessary).

e.g. `gpu: DisplayTransfer: a less amazing algorithm for flip` tells someone literally nothing about *why* the change was made.

File: src/core/hw/gpu.cpp [hunk 1]
```diff
@@ -123,11 +142,45 @@ static void MemoryFill(const Regs::MemoryFillConfig& config) {
...
+        return;
+    }
+
+    if (config.output_width == 0) {
+        LOG_CRITICAL(HW_GPU, "zero output width");
+        return;
+    }
+
+    if (config.output_height == 0) {
+        LOG_CRITICAL(HW_GPU, "zero input height");
```
Comment (Dragios) [inline src/core/hw/gpu.cpp]: `"zero output height"`

Comment (wwylele) [inline src/core/hw/gpu.cpp]: addressed

**Comment by j-selby** [comment]
In reference to 8b950fe - that makes a load more sense then my hack. Sorry that I've evidently made that a bigger issue for you guys then it really should have been.

File: src/video_core/renderer_opengl/gl_rasterizer.cpp [hunk 2]
```diff
@@ -692,10 +692,7 @@ bool RasterizerOpenGL::AccelerateDisplayTransfer(const GPU::Regs::DisplayTransfe
     using PixelFormat = CachedSurface::PixelFormat;
     using SurfaceType = CachedSurface::SurfaceType;
 
-    if (config.is_texture_copy) {
-        // TODO(tfarley): Try to hardware accelerate this
-        return false;
-    }
+    DEBUG_ASSERT(config.is_texture_copy == 0);
```
Comment (bunnei) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: Will these asserts ever be hit?

File: src/core/hw/gpu.cpp [hunk 3]
```diff
@@ -171,207 +358,23 @@ inline void Write(u32 addr, const T data) {
...
-                            break;
-                        }
-                    }
-                }
-
-                LOG_TRACE(HW_GPU, "DisplayTriggerTransfer: 0x%08x bytes from 0x%08x(%ux%u)-> "
+            if (config.is_texture_copy) {
+                TextureCopy(config);
+                LOG_TRACE(
+                    HW_GPU,
```
Comment (bunnei) [inline src/core/hw/gpu.cpp]: nit: Move this to line above, and tab over to it for subsequent lines (e.g like the LOG_TRACE below)

File: src/core/hw/gpu.cpp [hunk 4]
```diff
@@ -81,8 +81,27 @@ MICROPROFILE_DEFINE(GPU_DisplayTransfer, "GPU", "DisplayTransfer", MP_RGB(100, 1
 MICROPROFILE_DEFINE(GPU_CmdlistProcessing, "GPU", "Cmdlist Processing", MP_RGB(100, 255, 100));
 
 static void MemoryFill(const Regs::MemoryFillConfig& config) {
-    u8* start = Memory::GetPhysicalPointer(config.GetStartAddress());
-    u8* end = Memory::GetPhysicalPointer(config.GetEndAddress());
+    PAddr start_addr = config.GetStartAddress();
+    PAddr end_addr = config.GetEndAddress();
+
+    // TODO: do hwtest with these cases
```
Comment (bunnei) [inline src/core/hw/gpu.cpp]: Are you planning to HW test this before this gets merged, or is this just a note for someone in the future?

File: src/core/hw/gpu.cpp [hunk 5]
```diff
@@ -123,11 +142,45 @@ static void MemoryFill(const Regs::MemoryFillConfig& config) {
 }
 
 static void DisplayTransfer(const Regs::DisplayTransferConfig& config) {
+    PAddr src_addr = config.GetPhysicalInputAddress();
+    PAddr dst_addr = config.GetPhysicalOutputAddress();
+
+    // TODO: do hwtest with these cases
```
Comment (bunnei) [inline src/core/hw/gpu.cpp]: ditto

File: src/core/hw/gpu.cpp [hunk 6]
```diff
@@ -81,8 +81,27 @@ MICROPROFILE_DEFINE(GPU_DisplayTransfer, "GPU", "DisplayTransfer", MP_RGB(100, 1
 MICROPROFILE_DEFINE(GPU_CmdlistProcessing, "GPU", "Cmdlist Processing", MP_RGB(100, 255, 100));
 
 static void MemoryFill(const Regs::MemoryFillConfig& config) {
-    u8* start = Memory::GetPhysicalPointer(config.GetStartAddress());
-    u8* end = Memory::GetPhysicalPointer(config.GetEndAddress());
+    PAddr start_addr = config.GetStartAddress();
```
Comment (bunnei) [inline src/core/hw/gpu.cpp]: These can be `const`

File: src/core/hw/gpu.cpp [hunk 7]
```diff
@@ -123,11 +142,45 @@ static void MemoryFill(const Regs::MemoryFillConfig& config) {
 }
 
 static void DisplayTransfer(const Regs::DisplayTransferConfig& config) {
+    PAddr src_addr = config.GetPhysicalInputAddress();
```
Comment (bunnei) [inline src/core/hw/gpu.cpp]: These can be `const`

File: src/core/hw/gpu.cpp [hunk 8]
```diff
@@ -262,11 +315,40 @@ static void DisplayTransfer(const Regs::DisplayTransferConfig& config) {
 }
 
 static void TextureCopy(const Regs::DisplayTransferConfig& config) {
+    PAddr src_addr = config.GetPhysicalInputAddress();
+    PAddr dst_addr = config.GetPhysicalOutputAddress();
+
+    // TODO: do hwtest with these cases
```
Comment (bunnei) [inline src/core/hw/gpu.cpp]: ditto

File: src/core/hw/gpu.cpp [hunk 9]
```diff
@@ -262,11 +315,40 @@ static void DisplayTransfer(const Regs::DisplayTransferConfig& config) {
 }
 
 static void TextureCopy(const Regs::DisplayTransferConfig& config) {
+    PAddr src_addr = config.GetPhysicalInputAddress();
```
Comment (bunnei) [inline src/core/hw/gpu.cpp]: These can be `const`

File: src/core/hw/gpu.cpp [hunk 10]
```diff
@@ -416,10 +416,14 @@ inline void Write(u32 addr, const T data) {
             LOG_TRACE(HW_GPU, "MemoryFill from 0x%08x to 0x%08x", config.GetStartAddress(),
                       config.GetEndAddress());
 
-            if (!is_second_filler) {
-                GSP_GPU::SignalInterrupt(GSP_GPU::InterruptId::PSC0);
-            } else {
-                GSP_GPU::SignalInterrupt(GSP_GPU::InterruptId::PSC1);
+            // It seems that it won't signal interrupt if "address_start" is zero.
+            // TODO: hwtest this
```
Comment (wwylele) [inline src/core/hw/gpu.cpp]: For all these TODO hwtest I am not planning to do before merging. I am not so familiar with GPU, so want to leave them to others.

File: src/video_core/renderer_opengl/gl_rasterizer.cpp (hunk 2 above)
Comment (wwylele) [inline src/video_core/renderer_opengl/gl_rasterizer.cpp]: no, it's just an indicator how the function should be called. If it is redundant I will remove it.

**Comment by wwylele** [comment]
Addressed comments above.

**Comment by bunnei** [comment]
LGTM

**Pull request merged**